src/lib/optimizer/strategy/column_pruning_rule.cpp

```diff
@@ -113,42 +115,40 @@ ExpressionUnorderedSet ColumnPruningRule::_collect_actually_used_columns(const s
 
 void ColumnPruningRule::_prune_columns_from_leaves(const std::shared_ptr<AbstractLQPNode>& lqp,
                                                    const ExpressionUnorderedSet& referenced_columns) {
+  std::vector<std::shared_ptr<AbstractLQPNode>> leafs;
+
   // Collect all parents of leaves and on which input side their leave is
   // (if a node has two leaves as inputs, it will be collected twice)
   auto leaf_parents = std::vector<std::pair<std::shared_ptr<AbstractLQPNode>, LQPInputSide>>{};
   visit_lqp(lqp, [&](auto& node) {
-    for (const auto input_side : {LQPInputSide::Left, LQPInputSide::Right}) {
-      const auto input = node->input(input_side);
-      if (input && input->input_count() == 0) leaf_parents.emplace_back(node, input_side);
+    // Early out for non-leafs
+    if (node->input_count() != 0) {
+      return LQPVisitation::VisitInputs;
     }
 
-    // Do not visit the ProjectionNode we may have just inserted, that would lead to infinite recursion
-    return LQPVisitation::VisitInputs;
-  });
-
-  // Insert ProjectionNodes that prune unused columns between the leaves and their parents
-  for (const auto& parent_and_leaf_input_side : leaf_parents) {
-    const auto& parent = parent_and_leaf_input_side.first;
-    const auto& leaf_input_side = parent_and_leaf_input_side.second;
-    const auto leaf = parent->input(leaf_input_side);
-
-    // Collect all columns from the leaf that are actually referenced
-    auto referenced_leaf_columns = std::vector<std::shared_ptr<AbstractExpression>>{};
-    for (const auto& expression : leaf->column_expressions()) {
-      if (referenced_columns.find(expression) != referenced_columns.end()) {
-        referenced_leaf_columns.emplace_back(expression);
+    auto pruned_column_ids = std::vector<ColumnID>{};
+    for (const auto& expression : node->column_expressions()) {
+      if (referenced_columns.find(expression) == referenced_columns.end()) {
+        const auto column_expression = std::dynamic_pointer_cast<LQPColumnExpression>(expression);
+        pruned_column_ids.emplace_back(column_expression->column_reference.original_column_id());
       }
     }
 
-    if (leaf->column_expressions().size() == referenced_leaf_columns.size()) continue;
+    // We cannot create a Table without columns - since Chunks rely on their first column to determine their row count
```

Comment:
In which case would it be valid to have an LQP that does not return a single column? If you can‘t come up with an example, I‘d rather put an assert here and see if it ever gets triggered

Reply:
`SELECT 5 FROM t` would normally prune all columns from `t` and we'd loose the row count :(

Reply:
Good point. I think we should explicitly state that and added a comment.